File: templates/java-webmvc/subsystem/post/component/src/test/java/{{PACKAGE_NAME}}/post/PostServiceTest.java
package {{PACKAGE_NAME}}.post;

import {{PACKAGE_NAME}}.post.model.Post;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.BDDMockito.given;

@ExtendWith(SpringExtension.class)
class PostServiceTest {

  @Mock
  private PostRepository postRepository;

  @InjectMocks
  private PostService postService;

  @Test
  void findAll() {
    final List<Post> posts = MockData.posts;

    given(this.postRepository.findAll())
      .willReturn(posts);

    assertEquals(this.postService.findAll(), posts);
  }

  @Test
  void findByUserId() {
    final Integer userId = 1;
    final List<Post> posts = MockData.posts
      .stream()
      .filter(post -> post.id.equals(userId))
      .collect(Collectors.toList());

    given(this.postRepository.findByUserId(userId))
      .willReturn(posts);

    assertEquals(this.postService.findByUserId(userId), posts);
  }

  @Test
  void findById() {
    final Post post = MockData.posts.get(2);

    given(this.postRepository.findById(post.id))
      .willReturn(Optional.of(post));

    assertEquals(this.postService.findById(post.id), post);
  }

  @Test
  void notFound() {
    given(this.postRepository.findById(5))
      .willReturn(Optional.empty());

    assertThrows(ResponseStatusException.class, () -> this.postService.findById(5));
  }

}

File: templates/kotlin-webmvc/subsystem/post/component/src/test/kotlin/cobalt/post/PostServiceTest.kt
package cobalt.post

import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.springframework.http.HttpStatus
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.web.server.ResponseStatusException

@ExtendWith(SpringExtension::class)
internal class PostServiceTest {

  @Mock
  lateinit var postRepository: PostRepository

  @InjectMocks
  lateinit var postService: PostService

  @Test
  fun findAll() {
    val posts = MockData.posts

    given(this.postRepository.findAll())
      .willReturn(posts)

    assertEquals(this.postService.findAll(), posts)
  }

  @Test
  fun findByUserId() {
    val userId = 1
    val posts = MockData.posts.filter { it.userId == userId }

    given(this.postRepository.findByUserId(userId))
      .willReturn(posts)

    assertEquals(this.postService.findByUserId(userId), posts)
  }

  @Test
  fun findById() {
    val post = MockData.posts[2]

    given(this.postRepository.findById(post.id))
      .willReturn(post)

    assertEquals(this.postService.findById(post.id), post)
  }

  @Test
  fun notFound() {
    given(this.postRepository.findById(5))
      .willReturn(null)

    assertThrows(ResponseStatusException(HttpStatus.NOT_FOUND)::class.java) { this.postService.findById(5) }
  }

}